springql/src/config.rs

```diff
@@ -0,0 +1,196 @@
+pub struct SpringSourceReaderConfigDeserialize {
+    pub net_connect_timeout_msec: u32,
+    pub net_read_timeout_msec: u32,
+
+    pub can_read_timeout_msec: u32,
+}
+
+#[allow(missing_docs)]
+#[derive(Copy, Clone, Eq, PartialEq, Debug, Deserialize)]
+#[serde(remote = "SpringSinkWriterConfig")]
+pub struct SpringSinkWriterConfigDeserialize {
+    pub net_connect_timeout_msec: u32,
+    pub net_write_timeout_msec: u32,
+}
+
+impl SpringConfigDeserialize {
+    /// # Failures
+    ///
+    /// - [SpringError::InvalidConfig](crate::api::error::SpringError::InvalidConfig) when:
+    ///   - `overwrite_config_toml` includes invalid key and/or value.
+    /// - [SpringError::InvalidFormat](crate::api::error::SpringError::InvalidFormat) when:
+    ///   - `overwrite_config_toml` is not valid as TOML.
+    pub fn load(overwrite_config_toml: &str) -> Result<SpringConfig> {
+        let default_conf = config::Config::builder()
+            .add_source(config::File::from_str(
+                SPRING_CONFIG_DEFAULT,
+                config::FileFormat::Toml,
+            ))
+            .build()
+            .expect("SPRING_CONFIG_DEFAULT is in wrong format");
+
+        let c = config::Config::builder()
+            .add_source(default_conf)
+            .add_source(config::File::from_str(
+                overwrite_config_toml,
+                config::FileFormat::Toml,
+            ))
+            .build()
+            .map_err(|e| SpringError::InvalidFormat {
+                s: overwrite_config_toml.to_string(),
+                source: e.into(),
+            })?;
+
+        let de = c
+            .try_deserialize::<'_, Self>()
+            .map_err(|e| SpringError::InvalidConfig { source: e.into() })?;
+
+        Ok(SpringConfig {
+            worker: de.worker,
+            memory: de.memory,
+            web_console: de.web_console,
+            source_reader: de.source_reader,
+            sink_writer: de.sink_writer,
+        })
+    }
+
+    /// Configuration by TOML format string.
+    ///
+    /// # Parameters
+    ///
+    /// - `overwrite_config_toml`: TOML format configuration to overwrite default. See `SPRING_CONFIG_DEFAULT` in [spring_config.rs](https://github.com/SpringQL/SpringQL/tree/main/springql-core/src/api/spring_config.rs) for full-set default configuration.
+    ///
+    /// # Failures
+    ///
+    /// - [SpringError::InvalidConfig](crate::api::error::SpringError::InvalidConfig) when:
+    ///   - `overwrite_config_toml` includes invalid key and/or value.
+    /// - [SpringError::InvalidFormat](crate::api::error::SpringError::InvalidFormat) when:
+    ///   - `overwrite_config_toml` is not valid as TOML.
+    pub fn from_toml(overwrite_config_toml: &str) -> Result<SpringConfig> {
+        Self::load(overwrite_config_toml)
+    }
+}
+
+/// deserialize configration from toml
+pub fn config_from_toml(toml: &str) -> Result<SpringConfig> {
+    SpringConfigDeserialize::from_toml(toml)
+}
```

Comment:
I like this better:

- springql-core has `Config`, which does not impl Default and not visible to `springql` crate users.
- springql crate implements `SpringConfig`, which impl Default and visible to `springql` crate users. Also, it implements `pub fn new(overwrite_config_toml: &str) -> Result<Self>` as before.

- For what reason is the new API required?
- Why is this a free function (not inside `impl

Reply:
rust can not `impl` outside of crate.

```
   Compiling springql v0.15.0 (/home/kazuk/SpringQL/springql)
error[E0116]: cannot define inherent `impl` for a type outside of the crate where the type is defined
   --> springql/src/config.rs:193:1
    |
193 | / impl SpringConfig {
194 | |     pub fn new(toml: &str) -> Result<SpringConfig> {
195 | |         SpringConfigDeserialize::from_toml(toml)
196 | |     }
197 | | }
    | |_^ impl for type defined outside of crate.
    |
    = note: define and implement a trait or new type instead

For more information about this error, try `rustc --explain E0116`.
```

Reply:
I know it. So I claim that we move `SpringConfig` from springql-core into `springql` crate and then impl Default there.

Reply:
turn `SpringConfig` definition `struct` to` trait`?